compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp

```diff
@@ -528,6 +528,171 @@ wrapOpInDispatchRegion(RewriterBase &rewriter, Operation *op) {
   return newRegionOp;
 }
 
+FailureOr<Operation *> hoistOutOfDispatch(RewriterBase &rewriter,
+                                          Operation *op) {
+  assert(op && !isNonNullAndOutsideDispatch(op) &&
+         "op expected to be in a dispatch");
+
+  // Step 1: Clone the op outside of the dispatch region.
+
+  OpBuilder::InsertionGuard g(rewriter);
+  auto dispatchRegionOp = op->getParentOfType<DispatchRegionOp>();
+
+  // If all operands of the `op` come from outside the dispatch, then the op can
+  // be hoisted out before the dispatch region. Otherwise, the op can be hoisted
+  // out below the dispatch if the only users of the op are the dispatch return.
```

Comment:
> Does "hoist" refer to moving ops before? I thought It was moving them above and "sink" would refer to moving them after.

Hoist refers to pulling an operation out of a region that it is nested in. My understanding of the term is that it is not specific to being before or after.

> Also, if this is in RegionOpUtils, it might make sense to separate this into 2 functions and expose a hoist/move above & sink/move below version. That may simplify this a bit because it seems like it is trying to do 2 things simultaneously

I kind of like having it as a single function, since hoisting before or after both need to do a lot of the same things. The main difference is where the insertion point is. Keeping the logic to decide whether to hoist before or after inside the function means that callers don't have to worry about dominance issues when using the function.